flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -140,10 +164,19 @@ class ExprLowering {
   mlir::Type getSomeKindInteger() { return builder.getIndexType(); }
 
   template <typename OpTy, typename A>
-  mlir::Value createBinaryOp(const A &ex, mlir::Value lhs, mlir::Value rhs) {
-    assert(lhs && rhs && "argument did not lower");
-    auto x = builder.create<OpTy>(getLoc(), lhs, rhs);
-    return x.getResult();
+  mlir::Value createBinaryOp(const A &ex, const Fortran::lower::ExValue &lhs,
+                             const Fortran::lower::ExValue &rhs) {
+    return std::visit(Fortran::common::visitors{
+                          [&](const fir::UnboxedValue &lhs,
+                              const fir::UnboxedValue &rhs) -> mlir::Value {
+                            assert(lhs && rhs && "argument did not lower");
+                            return builder.create<OpTy>(getLoc(), lhs, rhs);
+                          },
+                          [](const auto &lhs, const auto &rhs) -> mlir::Value {
+                            // binary ops can appear in array contexts
+                            TODO();
+                          }},
+                      lhs, rhs);
```

Comment:
I'm a bit of two minds on the exact structure of these dictionaries, as I said before. There are all sorts of ways it could be implemented, and this PR splits the difference to a degree. Each approach would have all the same advantages and disadvantages as anywhere else.

- The parametric polymorphism approach (using `variant`) gives type safety and autogenerates an internal enum distinguishing each box type.
- Another approach could be to use "the most generalized" box and define our own distinguishing enum values.
- There are many other ways to bundle the data as well.

To me, making the data structure easy to use, easy to extend, and efficient top the priorities.

Each node of the tree must be visited. Each node in the tree must be handled. The nodes in the tree will (mostly) have executable semantics. That representation must be exposed to the parent node in the tree. It's hard to see where a generalized recursive descent over the expression tree would not be correct.

On the other hand, it can be argued that the tree itself is not conducive to easy transformation by its very structure. One facility for dealing with that already exists in the PFT, which can be used to rearrange the tree structure to impose simplifying preconditions. Another facility would be to have special-purpose recursive functions to preprocess expressions as needed. But neither of those would change the fundamental nature of lowering a tree-like data structure.